### PR TITLE
feat: Added support for GitHub app

### LINKED ISSUE #271: Allow running atlantis with github app instead of a user token
## Is your request related to a problem? Please describe.

Since we want to use atlantis in a GitHub organization, I want to have this environment variables and secrets in my atlantis container without any github user tokens.:

```terraform
  custom_environment_variables = [
    {
      name  = "ATLANTIS_GH_APP_ID"
      value = local.github_app_id
    },
    {
      "name" : "ATLANTIS_WRITE_GIT_CREDS",
      "value" : "true",
    }
  ]

  custom_environment_secrets = [
    {
      name      = "ATLANTIS_GH_APP_KEY"
      valueFrom = local.atlantis_github_app_key_ssm_parameter_name
    }
  ]
```
The module currently depends on the presence of `var.atlantis_github_user_token`, which is used as a conditional in the creation of some resources and that brings a problem:
- Without the atlantis_github_user_token, although there may be other secrets in the environment(webhook_secret. github app key secret etc.), the tasks can not get their value.
https://github.com/terraform-aws-modules/terraform-aws-atlantis/blob/fab764c06444a0e72deb6c3cf92bdb4fbc31ba4b/main.tf#L567

I worked around this issue by modifying the module by adding some custom variables.

## Describe the solution you'd like.

I would like to create the atlantis module with a simple parameter like `use_github_bot = true` which will take care of setting up all the necessary environment variables/parameter store secrets for me.

### PR DESCRIPTION
**Pull request by abalcobia**
## Description
Allow users to utilize the GitHub App instead of a personal access token to execute Atlantis.

## Motivation and Context
Enable the use of a GitHub App with Atlantis. In this PR, four variables were added to the module and their dependencies: 
- `var.atlantis_github_app_key` 
- `var.atlantis_github_app_key_ssm_parameter_name`
- `var.atlantis_write_git_creds`
- `var.atlantis_github_app_id`

Note: It is not necessary to create a specific variable for the webhook secret in the GitHub App, the existing one can be used.

- fixes #271

ref:https://www.runatlantis.io/docs/access-credentials.html#generating-an-access-token (GitHub App section)

## Breaking Changes
None

## How Has This Been Tested?
- [x] I have executed `pre-commit run -a` on my pull request

Initially, a github app was created and consequently a private key and a secret webhook were associated with 
it. Posteriorly, a test repo was used with the inputs (atlantis_github_app_id, atlantis_github_app_key & atlantis_github_webhook_secret) from the github app to guarantee the communication between Atlantis and the GitHub app at the level of webhooks and comments and it was possible to guarantee this communication with success as well as the provisioning of resources.

### CODE REVIEW

File: main.tf
Comment (bryantbiggs) [inline main.tf:18]: I believe we can wrap all these in this (untested):
```suggestion
  has_secrets = try(coalesce(var.atlantis_gitlab_user_token, var.atlantis_github_user_token, var.atlantis_github_app_key, var.atlantis_bitbucket_user_token) != "", false)
```

Comment (bryantbiggs) [reply main.tf:18]: or perhaps better yet:

```suggestion
  has_secrets = compact([var.atlantis_gitlab_user_token, var.atlantis_github_user_token, var.atlantis_github_app_key, var.atlantis_bitbucket_user_token]) > 0
```

File: main.tf
Comment (bryantbiggs) [inline main.tf:587]:
```suggestion
  count = local.has_secrets ? 1 : 0
```

File: main.tf
Comment (bryantbiggs) [inline main.tf:20]: oof, I'll have to check this out locally to see if I can follow or make sense of this. its getting a little too unwieldy

**Comment by bryantbiggs** [comment]
Thanks for the PR @abalcobia  - since GitHub apps are more secure and the recommended way to provide access, lets update the `github-complete` to use this functionality and provide any docs that users would need to get this setup to use or validate

File: main.tf
Comment (abalcobia) [reply main.tf]: I tested both suggestions, but only the first one went well. I made a commit with the proposed change, ran it locally and it worked fine with my test repo.

File: main.tf
Comment (abalcobia) [reply main.tf]: I tested this suggestion and it ran fine. The change is included in the "code improvements" commit.

**Comment by abalcobia** [comment]
Regarding the updates and the docs on examples, I will work on that in the next days.

**Comment by calexandre** [comment]
@bryantbiggs  @antonbabenko I've helped @abalcobia with the latest changes. Can you guys review please?

File: examples/github-complete/main.tf
Comment (bryantbiggs) [inline examples/github-complete/main.tf:78]: Lets remove this - I don't think it would be correct to have both user based auth and app based

File: examples/github-complete/terraform.tfvars.sample
Comment (bryantbiggs) [inline examples/github-complete/terraform.tfvars.sample:4]: user and token would be removed here as well

File: examples/github-complete/variables.tf
Comment (bryantbiggs) [inline examples/github-complete/variables.tf:25]: and user and token variables here

**Comment by calexandre** [comment]
Thanks for the review @bryantbiggs! we'll make the requested changes asap 👍

**Comment by abalcobia** [comment]
Hello @bryantbiggs! Just a question to clarify the approach, is it only supposed to remove the user and token variables, and respective dependencies, in the folder examples or throughout the entire repo?

**Comment by bryantbiggs** [comment]
@abalcobia - correct, just remove those from the example. It doesn't make a lot of sense to have two authentication routes for one example, I don't even know if this works, so lets just show the "preferred" way which is to use the GitHub app authentication route in the example. Does that make sense? The user/token authentication route will still exist in the module itself, we're just changing the example to show using the GitHub app route.

**Comment by abalcobia** [comment]
@bryantbiggs - yes, makes sense, thanks. We get the idea and throughout the day we will update the README.md file only in the examples folder.

**Comment by calexandre** [comment]
@bryantbiggs  we've included instructions on how to generate a github app. We've also added a note stating that using GitHub PATs is no longer recommended (despite being still supported).

Let us know if you need any more changes.

**Comment by calexandre** [comment]
@bryantbiggs we've updated the PR with the latest changes from the `master`.

Can we get this merged on the master, or is something missing?

**Comment by abalcobia** [comment]
Hello @bryantbiggs @antonbabenko 

If you feel that any more changes need to be made please let us know. 

Thanks.

**Comment by bryantbiggs** [comment]
I will test this today

**Comment by calexandre** [comment]
@bryantbiggs  sorry to bother, but do you have any updates on this? We've updated the branch to the latest commit on master...

**Comment by matthiassb** [comment]
Commenting to keep open

**Comment by calexandre** [comment]
rebased and squashed the commit history

**Comment by bryantbiggs** [comment]
apologies for the delay - just getting back around to this. I'm a bit confused: what is the benefit here for GitHub app support if GitHub PAT is still required?

**Comment by calexandre** [comment]
Well, it depends if you want to keep supporting PAT based installations...
Do you want us to remove all PAT references?

**Comment by bryantbiggs** [comment]
I don't think thats the way to look at it - there are two avenues:
1. User elects to use GitHub PAT (and only PAT) 
2. User elects to use GitHub App (and only App - no PAT)

I believe both can be supported for now to avoid a breaking change, but this current implementation does not satisfy point 2 above which makes the current implementation a bit confusing. Why use a GitHub app if a PAT is still required?

**Comment by calexandre** [comment]
> I don't think thats the way to look at it - there are two avenues:
> 
> 1. User elects to use GitHub PAT (and only PAT)
> 2. User elects to use GitHub App (and only App - no PAT)
> 
> I believe both can be supported for now to avoid a breaking change, but this current implementation does not satisfy point 2 above which makes the current implementation a bit confusing. Why use a GitHub app if a PAT is still required?

I'm confused...All PAT related variables are optional..
Can you please request a change on where the current implementation does not satisfy the requirements? Maybe it is easier to understand what you mean...

File: examples/github-complete/terraform.tfvars.sample
Comment (bryantbiggs) [reply examples/github-complete/terraform.tfvars.sample:4]: @calexandre

**Comment by bryantbiggs** [comment]
@calexandre simply try deploying the `github-complete` example, it does not work as defined in this PR currently. You can see why it won't work because a PAT is still required for the webhook https://github.com/terraform-aws-modules/terraform-aws-atlantis/blob/master/modules/github-repository-webhook/main.tf#L2-L4

**Comment by calexandre** [comment]
Oh, that must have slipped on a previous rebase...
I will look into that in a moment!

Thank you for pointing it out 👍

**Comment by calexandre** [comment]
Just note that it is not possible to configure the github provider using both approaches, you either need to choose the PAT of the APP approach, or use two providers...

**Comment by bryantbiggs** [comment]
> Just note that it is not possible to configure the github provider using both approaches, you either need to choose the PAT of the APP approach, or use two providers...

Is it possible to set the variables to `null` as default - I was going to try this but the example wasn't working to start so didn't get to it. Might have to be part of a larger refactor and externalize the provider like usual

**Comment by calexandre** [comment]
@bryantbiggs I re-added the `github_token` variable on all modules that require `github-repository-webhook` to make this work for both scenarios.

Please note that only Atlantis should benefit from the github app auth method as it is more suited for this type applications.

It is debatable whether the `github-repository-webhook` should use PAT or APP authentication; regardless of which method is preferred I believe it is outside of the scope of this PR, since it would have to be a different APP with a different set of permissions.

Note that Atlantis default/recommended application permissions don't involve any kind of repository settings interaction, which need to be changed in order to make it work to create webhooks.

**Comment by calexandre** [comment]
> > Just note that it is not possible to configure the github provider using both approaches, you either need to choose the PAT of the APP approach, or use two providers...
> 
> Is it possible to set the variables to `null` as default - I was going to try this but the example wasn't working to start so didn't get to it. Might have to be part of a larger refactor and externalize the provider like usual

I was not able to setup the github provider to work with both methods in the same provider. The only way is to setup a second provider alias and use the APP auth on that one.
You are right, the correct way would be to take the provider out of the module so that the invoker controls the provider authentication.

File: examples/github-complete/terraform.tfvars.sample
Comment (calexandre) [reply examples/github-complete/terraform.tfvars.sample:4]: I cannot remove this because it would break the repository webhook creation.

File: examples/github-complete/variables.tf
Comment (calexandre) [reply examples/github-complete/variables.tf]: I cannot remove this because it would break the repository webhook creation.

**Comment by calexandre** [comment]
@bryantbiggs  I also forgot to mention that with a GitHub App you no longer need to create a repository webhook for sending events... That part is configured on the GitHub App settings.

One suggestion to make the example a bit more elegant could be to only invoke the webhook module when the app authentication variables are not set...

**Comment by bryantbiggs** [comment]
if thats the case then lets remove the webhook from the example. the webhook is only there to send the payload to Atlantis when there are changes

**Comment by bryantbiggs** [comment]
(my bad - wrong button 🤣 )

**Comment by calexandre** [comment]
> if thats the case then lets remove the webhook from the example. the webhook is only there to send the payload to Atlantis when there are changes

In that case I will remove all variables and references to the `github-repository-webhook` module (including the module itself).
Is that okay?

**Comment by calexandre** [comment]
> > if thats the case then lets remove the webhook from the example. the webhook is only there to send the payload to Atlantis when there are changes
> 
> In that case I will remove all variables and references to the `github-repository-webhook` module (including the module itself). Is that okay?

@bryantbiggs could you please give more details on what you mean by removing the webhook from the example? You meant to remove all the webhook implementations, or just from the particular `github-complete` example while keeping the module, etc..?

On the other hand, we could also copy the existing example and name it to something like `github-complete-app-auth` and remove the webhook references from there. That way you keep both examples and preserve retrocompatibility.

Just let me know how you want to proceed so that we can get this merged :)

**Comment by bryantbiggs** [comment]
Just remove the webhook implementation from the `github-complete` example. We'll leave the sub-module in the project so that we don't break the existing implementation for users

**Comment by calexandre** [comment]
@bryantbiggs I've updated the `github-complete` example and removed the `github-repository-webhook` dependency from the example.

**Comment by calexandre** [comment]
@bryantbiggs  is there anything else missing for merging this PR?

**Comment by andyshinn** [comment]
Can we reopen this and get it merged?

**Comment by calexandre** [comment]
I've rebased the branch so that you can review.
@antonbabenko @dynamike  let me know if you want to make any additional changes.

File: examples/github-complete/README.md
Comment (dynamike) [inline examples/github-complete/README.md:9]: This option doesn't seem possible because as you've mentioned you need an exposed endpoint to be able to setup the GitHub App. How would one go about doing that? Should I - 
 * Spin up a version of the `github-complete` example without a valid `github_app_id`, `github_app_key`, and `github_webhook_secret`
 * Register the GitHub app via the exposed endpoint
 * Re-run terraform with the GitHub app values?

More detail here is needed if this is the preferred path.

File: examples/github-complete/README.md
Comment (dynamike) [inline examples/github-complete/README.md:10]: I also ran into a wall with this method. The Github docs on setting up a Github App  are very generic and doesn't explain how I should put values in for sections like `Identifying and authorizing users` or `Webhook URL` to work with Atlantis. So I'm left without a path forward to test this out.

File: examples/github-complete/README.md
Comment (dynamike) [inline examples/github-complete/README.md:20]: I think it makes sense to support both the Github App **and** the Personal Access Token (PAT) method in this module. While it's true you get fine grained access with Github Apps, there is a GitHub beta feature to allow you to created [PATs with fine grained access](https://github.blog/2022-10-18-introducing-fine-grained-personal-access-tokens-for-github/). So the benefit of a Github App vs a PAT become less clear.

**Review by dynamike — CHANGES_REQUESTED** [review]
I wasn't really able to test this feature out because of a difficulty in bootstrapping GitHub apps, which I describe in this review. Let me know what you think as I think this would be a handy feature.

File: examples/github-complete/README.md
Comment (calexandre) [reply examples/github-complete/README.md:20]: Note that despite the PAT's new features, you still require a valid user to create the PAT for. This means consuming an additional license for certain organizations and using an authentication mechanism whose purpose is for users and not apps/services.

Apart from fine grained access, there are multiple advantages by using a GitHub App, one of them being the extensive webhook configuration options, among other things.

Since Atlantis is an app, it fits the GitHub App's purpose.

File: examples/github-complete/README.md
Comment (calexandre) [reply examples/github-complete/README.md:9]: I understand your pain... it's not very easy to bootstrap Atlantis using the official documentation.

What the Atlantis team did, was to try to ease the process of setting up the GitHub App by using the GitHub App manifest configuration method - this 'manifest' method, involves the Atlantis runtime to provide all the requirements for GitHub to create the app on Atlantis behalf. Of course, this raises security concerns: 
- GitHub needs to trust the exposed endpoint (this means not only being exposed, but needs to have a valid SSL certificate)
- You need to expose your app to the internet and follow the 'one-time' setup procedure to get the required secrets `github_app_id`, `github_app_key`, and `github_webhook_secret`.
- This method not only provides you the secrets, but also configures the GitHub App with the required permissions as depicted by the Atlantis team.

**TL;DR:** The steps you mentioned are correct. That's what you need to do in order to setup the GitHub App according to Atlantis official documentation.

File: examples/github-complete/README.md
Comment (calexandre) [reply examples/github-complete/README.md:10]: I understand the confusion. This is an advanced setup which requires the user to be comfortable with GitHub Apps, hence the above recommended path depicted by the Atlantis team.

Let me try and make this clearer:

- You create a GitHub App and give it a name - that name must be unique across the world (you can change it later).
- Provide a valid Homepage URL (this can be the atlantis server url, for instance `https://atlantis.mydomain.com`)
- Provide a valid Webhook URL. The Atlantis webhook server path is located by default at `https://atlantis.mydomain.com/events`
- Generate a Webhook Secret - this is used for Atlantis to trust the deliveries. This is your `github_webhook_secret`.
- Generate a Private Key - this is your `github_app_key`
- On the App's settings page (at the top) you find the App ID. This is your `github_app_id`
- On the **Permissions & Events** you need to setup all the permissions and events according to [Atlantis documentation](https://www.runatlantis.io/docs/access-credentials.html#permissions)

Now you need to Install the App on your organization.

A self-provisioned GitHub App usually has two parts: the App and the Installation.
- The App part is the first step and its where you setup all the requirements, such as authentication, webhook, permissions, etc...
- The Installation part is where you add the created App to an organization/personal-account. It is on the installation page where you setup which repositories the application can access and receive events from.

File: examples/github-complete/README.md
Comment (calexandre) [reply examples/github-complete/README.md:20]: I'd like to add that if your organization has an SSO, using a PAT is a bit cumbersome.

File: examples/github-complete/README.md
Comment (bryantbiggs) [reply examples/github-complete/README.md:10]: All of this is great info to add to the module docs for users

**Review by dynamike — COMMENTED** [review]
I was able to get this working using the method in the Atlantis documentation, but I had to make some changes to the github-complete example for it to work correctly. You can see the changes I made over at https://github.com/nosportugal/terraform-aws-atlantis/compare/github-app-support...dynamike:terraform-aws-atlantis:github-app-support?expand=1 as I don't really have a way to incorporate those changes in this PR.

**Comment by calexandre** [comment]
> I was able to get this working using the method in the Atlantis documentation, but I had to make some changes to the github-complete example for it to work correctly. You can see the changes I made over at https://github.com/nosportugal/terraform-aws-atlantis/compare/github-app-support...dynamike:terraform-aws-atlantis:github-app-support?expand=1 as I don't really have a way to incorporate those changes in this PR.

Looks great! Just waiting for the review on the branch...

**Review by dynamike — APPROVED** [review]
@antonbabenko I think this is ready to go. I've confirmed there aren't breaking changes which means using a traditional webhook or a Github App both work. The github-complete example is also functional using the Github App method. 

The one thing that I would note is that [ATLANTIS_WRITE_GIT_CREDS](https://www.runatlantis.io/docs/server-configuration.html#write-git-creds) is now defaulted to true because it is required if you are using the GitHub App method, but this will also enable the flag for folks not using the Github App.

File: examples/github-complete/README.md
Comment (antonbabenko) [inline examples/github-complete/README.md:9]:
```suggestion
To run this code you need to copy `terraform.tfvars.sample` into `terraform.tfvars` and update the values locally or specify them using environment variables (`TF_VAR_github_app_id=xxx`, `TF_VAR_github_owner=xxx`, etc.). Ensure that `bootstrap_github_app` is `true`.  Once ready, execute:
```

**Review by antonbabenko — APPROVED** [review]

**Comment by dynamike** [comment]
@calexandre Fixes for the tflint warnings are in https://github.com/nosportugal/terraform-aws-atlantis/pull/7

**Comment by dynamike** [comment]
@antonbabenko I think this one is ready to merge when you're able. Thanks!

**Comment by antonbabenko** [comment]
This PR is included in version 3.23.0 :tada:

**Comment by andyshinn** [comment]
Thanks for getting this over the finish line everyone!

**Comment by abalcobia** [comment]
Thank you to everyone involved for the effort made to get the PR integrated!

**Pull request merged**
